src/claim_canister/src/lib.rs

```diff
@@ -0,0 +1,265 @@
+            claims_ref.insert(claim_id, claim);
+            Ok(())
+        } else if claim.status == ClaimStatus::Approved {
+            Err(ClaimError::AlreadyApproved)
+        } else if claim.status == ClaimStatus::Executed {
+            Err(ClaimError::AlreadyExecuted)
+        } else {
+            Err(ClaimError::InvalidStatus)
+        }
+    })
+}
+
+#[ic_cdk::update]
+pub async fn execute_claim(claim_id: u64) -> Result<(), ClaimError> {
+    let claim = CLAIMS.with(|claims| claims.borrow().get(&claim_id).ok_or(ClaimError::NotFound))?;
+
+    if claim.status != ClaimStatus::Approved {
+        return Err(ClaimError::NotApproved);
+    }
+
+    let current_time = ic_cdk::api::time();
+    let approved_time = claim.approved_at.ok_or(ClaimError::NotApproved)?;
+
+    if current_time < approved_time + TIMELOCK_DURATION {
+        return Err(ClaimError::TimelockNotExpired);
+    }
+
+    let mut executing_marked = false;
+    CLAIMS.with(|claims| {
+        let mut claims_ref = claims.borrow_mut();
+        if let Some(mut c) = claims_ref.get(&claim_id) {
+            if c.status == ClaimStatus::Approved {
+                c.status = ClaimStatus::Executing;
+                claims_ref.insert(claim_id, c);
+                executing_marked = true;
+            }
+        }
+    });
+
+    if !executing_marked {
+        return Err(ClaimError::AlreadyExecuting);
+    }
+
+    let slash_result: Result<(Result<(), String>,), _> = call(
+        claim.pool_canister_id,
+        "slash",
+        (claim.receiver, claim.amount.clone()),
+    )
+    .await;
+
+    if let Ok((Ok(()),)) = slash_result {
+        CLAIMS.with(|claims| {
+            let mut claims_ref = claims.borrow_mut();
+            if let Some(mut updated_claim) = claims_ref.get(&claim_id) {
+                updated_claim.status = ClaimStatus::Executed;
+                claims_ref.insert(claim_id, updated_claim);
+            }
+        });
+        Ok(())
+    } else {
+        CLAIMS.with(|claims| {
+            let mut claims_ref = claims.borrow_mut();
+            if let Some(mut updated_claim) = claims_ref.get(&claim_id) {
+                updated_claim.status = ClaimStatus::Approved;
+                claims_ref.insert(claim_id, updated_claim);
+            }
+        });
+        Err(ClaimError::PoolCallFailed)
+    }
+}
+
+#[ic_cdk::query]
+pub fn get_claim(claim_id: u64) -> Option<ClaimInfo> {
+    CLAIMS.with(|claims| {
+        claims.borrow().get(&claim_id).map(|claim| {
+            let current_time = ic_cdk::api::time();
+            let can_execute = if claim.status == ClaimStatus::Approved {
+                if let Some(approved_time) = claim.approved_at {
+                    current_time >= approved_time + TIMELOCK_DURATION
+                } else {
+                    false
+                }
+            } else {
+                false
+            };
+
+            let time_until_execution = if claim.status == ClaimStatus::Approved {
```

Comment:
unite can_execute and time_until_execution fetching under the same ifs, as they are identical.